lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -1972,6 +2014,106 @@ void PullbackCloner::Implementation::emitZeroDerivativesForNonvariedResult(
+  auto *optArgBuf = builder.createAllocStack(pbLoc, optionalOfWrappedTanType);
+  if (optionalOfWrappedTanType.isLoadableOrOpaque(builder.getFunction())) {
+    // %enum = enum $Optional<T.TangentVector>, #Optional.some!enumelt,
+    //         %wrappedAdjoint : $T
+    auto *enumInst = builder.createEnum(pbLoc, wrappedAdjoint, someEltDecl,
+                                        optionalOfWrappedTanType);
+    // store %enum to %optArgBuf
+    builder.emitStoreValueOperation(pbLoc, enumInst, optArgBuf,
+                                    StoreOwnershipQualifier::Trivial);
+  } else {
+    // %enumAddr = init_enum_data_addr %optArgBuf $Optional<T.TangentVector>,
+    //                                 #Optional.some!enumelt
+    auto *enumAddr = builder.createInitEnumDataAddr(
+        pbLoc, optArgBuf, someEltDecl, wrappedTanType.getAddressType());
+    // copy_addr %wrappedAdjoint to [initialization] %enumAddr
+    builder.createCopyAddr(pbLoc, wrappedAdjoint, enumAddr, IsNotTake,
+                           IsInitialization);
+    // inject_enum_addr %optArgBuf : $*Optional<T.TangentVector>,
+    //                  #Optional.some!enumelt
+    builder.createInjectEnumAddr(pbLoc, optArgBuf, someEltDecl);
+  }
+
+  // Apply `Optional<T>.TangentVector.init`.
+  SILOptFunctionBuilder fb(getContext().getTransform());
+  // %init_fn = function_ref @Optional<T>.TangentVector.init
+  auto *initFn = fb.getOrCreateFunction(pbLoc, SILDeclRef(constructorDecl),
+                                        NotForDefinition);
+  auto *initFnRef = builder.createFunctionRef(pbLoc, initFn);
+  auto *diffProto =
+      builder.getASTContext().getProtocol(KnownProtocolKind::Differentiable);
+  auto *swiftModule = getModule().getSwiftModule();
+  auto diffConf =
+      swiftModule->lookupConformance(wrappedType.getASTType(), diffProto);
+  assert(!diffConf.isInvalid() && "Missing conformance to `Differentiable`");
+  auto subMap = SubstitutionMap::get(
+      initFn->getLoweredFunctionType()->getSubstGenericSignature(),
+      ArrayRef<Type>(wrappedType.getASTType()), {diffConf});
+  // %metatype = metatype $Optional<T>.TangentVector.Type
+  auto metatypeType = CanMetatypeType::get(optionalTanTy.getASTType(),
+                                           MetatypeRepresentation::Thin);
+  auto metatypeSILType = SILType::getPrimitiveObjectType(metatypeType);
+  auto metatype = builder.createMetatype(pbLoc, metatypeSILType);
+  // apply %init_fn(%optTanAdjBuf, %optArgBuf, %metatype)
+  builder.createApply(pbLoc, initFnRef, subMap,
+                      {optTanAdjBuf, optArgBuf, metatype});
+  builder.emitDestroyAddr(pbLoc, optArgBuf);
+  builder.createDeallocStack(pbLoc, optArgBuf);
+
+  // Accumulate adjoint for the incoming `Optional` value.
+  addToAdjointBuffer(bb, optionalValue, optTanAdjBuf, pbLoc);
+  // TODO: Investigate why enabling this produces a crash for `Tracked<Float>?`
+  // differentiation.
+  // builder.emitDestroyAddr(pbLoc, optTanAdjBuf);
```

Comment:
Can we figure out what's going on with this before merging?  Why is it a crasher rather than an ownership verification error?

Reply:
> Why is it a crasher rather than an ownership verification error?

I have the same question.

Uncommenting this `builder.emitDestroyAddr(pbLoc, optTanAdjBuf)` causes the test to fail with an "extra release" error at runtime, which seems unexpected because `optTanAdjBuf` is a locally created `alloc_stack`.

```
[ RUN      ] OptionalDifferentiation.Let
stderr>>> Assertion failed: (getStrongExtraRefCount() >= dec && "releasing reference whose refcount is already zero"), function decrementStrongExtraRefCount, file /Users/danielzheng/swift-merge/swift/include/swift/Runtime/../../../stdlib/public/SwiftShims/RefCount.h, line 567.
stderr>>> CRASHED: SIGABRT
```

Interestingly, commenting `builder.emitDestroyAddr` doesn't appear to cause memory leaks when running the test (according to `valgrind`).

Reply:
Can you dump the generated raw SIL of both cases and let me take a look?

Reply:
Yes, thanks for offering to look.

[Here's a crash reproducer](https://gist.github.com/dan-zheng/f49f5e3def048c786c453e05f657d008), with SIL for both cases and full `-Xllvm debug-only=differentiation` output.

Reply:
Here's the issue.  You called `Optional.TangentVector.init(_:)` with argument `%107` but you had a `destroy_addr %107` later in the code.  This itself is incorrect: `%107` should be considered consumed after the call.

```swift
  %107 = alloc_stack $Optional<Tracked<Float>>    // users: %115, %114, %113, %110, %108
  %108 = init_enum_data_addr %107 : $*Optional<Tracked<Float>>, #Optional.some!enumelt // user: %109
  copy_addr %3 to [initialization] %108 : $*Tracked<Float> // id: %109
  inject_enum_addr %107 : $*Optional<Tracked<Float>>, #Optional.some!enumelt // id: %110
  // function_ref Optional<A>.TangentVector.init(_:)
  %111 = function_ref @$sSq16_DifferentiationAA14DifferentiableRzlE13TangentVectorVyADyx_GACQzSgcfC : $@convention(method) <τ_0_0 where τ_0_0 : Differentiable> (@in Optional<τ_0_0.TangentVector>, @thin Optional<τ_0_0>.TangentVector.Type) -> @out Optional<τ_0_0>.TangentVector // user: %113
  %112 = metatype $@thin Optional<Tracked<Float>>.TangentVector.Type // user: %113
  %113 = apply %111<Tracked<Float>>(%106, %107, %112) : $@convention(method) <τ_0_0 where τ_0_0 : Differentiable> (@in Optional<τ_0_0.TangentVector>, @thin Optional<τ_0_0>.TangentVector.Type) -> @out Optional<τ_0_0>.TangentVector
BAD! --->  destroy_addr %107 : $*Optional<Tracked<Float>>  // id: %114
  dealloc_stack %107 : $*Optional<Tracked<Float>> // id: %115
```

The `destroy_value %107` did not happen to crash because coincidentally `TangentVector.init(_:)` is simply wrapping `%107` in a `TangentVector` wrapper. When you `destroy_value %107`, it released the underlying value. Then when you `destroy_value %106`, which at this point still wraps `%107`'s value, it's a double-free.

Reply:
@gottesmm Any idea why the ownership verifier didn't catch this?
```console
%107 = alloc_stack $Optional<Tracked<Float>>
%113 = apply %111<Tracked<Float>>(%106, %107, %112) : $@convention(method) <τ_0_0 where τ_0_0 : Differentiable> (@in Optional<τ_0_0.TangentVector>, @thin Optional<τ_0_0>.TangentVector.Type) -> @out Optional<τ_0_0>.TangentVector
**Use after consume**: destroy_addr %107 : $*Optional<Tracked<Float>>  // id: %114
```

Reply:
Thanks. That makes sense, because `Optional<A>.TangentVector.init(_:)` consumes its `@in` parameter. Fixed.